### DISCUSSION TITLE
CarelinkFollower notifications only in Dev build.

### DISCUSSION THREAD

**Discussion by kukuchta** [NightscoutFoundation/xDrip#3671] [Bug reports] [answered]
I have started investigating why at some point xDrip stopped displaying the Carelink notifications (alerts) on the main chart.

It seems that this functionality works on the Dev builds, but not on the Release. Is it possible that minifying the app cuts out too much of the code? I see that the RecentData data structure, that should hold all the info and notifications that were downloaded and deserialized from Carelink is missing the key member, the NotificationHistory, and therefore does not allow the notifications to be displayed.

I'm not experienced in Java/Kotlin/Android studio, so I'm not sure how to fix this. Can any of you guys help and check this? @benceszasz ?

For reference, there are screenshots of the same period of time using dev and release builds:

![Screenshot_20240918_174220.png](https://github.com/user-attachments/assets/c2c6057d-3160-4405-a2f9-5924861e3fb4)

![Screenshot_2024-09-18-17-39-46-61_2dd2c1bcb8f659cd696bcba2072f751b.jpg](https://github.com/user-attachments/assets/292caac3-4275-48c1-9ca2-72ed2715a901)

**Comment by Navid200** [comment]
Can you post a screenshot of the classic status page from the version that works and the version that doesn't?

**Reply by kukuchta** [reply]
Sure. Previous screenshots were taken partially on the phone. 

This time I did everything on AVD and paying attention to do exactly same steps for both builds:
- clear xDrip data and cache on AVD
- uninstall old xDrip build from AVD
- set the Active Build Variant for "app" (fastDev or fastDebug)
- rebuild project
- run app on AVD
- set CarelinkFollower as data source
- enable Notifications setting
- login to Carelink
- make screenshots

I used most recent commit from master branch.

Result for **fastDebug**:
![debug1.png](https://github.com/user-attachments/assets/bf9bc9a9-8bb3-40af-a4f0-c31a9d88922e)

![debug2.png](https://github.com/user-attachments/assets/d74216b8-662f-44fb-b67e-1c33fdcda89f)

Result for **fastDev**:
![dev1.png](https://github.com/user-attachments/assets/bc28c146-bf8b-4298-bc49-947b49412748)

![dev2.png](https://github.com/user-attachments/assets/824246c0-8e83-44bd-b2c6-a34164323c3b)

**Reply by Navid200** [reply]
xDrip is open source as evidently you know.
We have a large group of people who use xDrip.  I am not talking about those who compile xDrip.
The main purpose of these forums is to help the users.  It also helps us figure out what is wrong with xDrip and fix it.

If you self compile xDrip and have problems, we can talk about it here.  It will be a completely different conversation.  
Why xDrip doesn't work will not be included in that conversation.  If you need help using xDrip, you should not be compiling it.

**Reply by kukuchta** [reply]
Okay, let me get this straight. 
I could define my problem as: **"Any realease of xDrip downloaded since this spring does not show the notification data on the screen."** Period.

I even started a discussion sometime ago, but no one ever answered.

Now that I have some free time, I cloned the repo and started investigating myself. I think I found some relevant clues so I wanted to contribute a bit. Im a C# dev, not java, so my knowledge is very limited here.

That said, whats up with the tone of your previous message? Did I offend you somehow by trying to solve this issue? Geeez...

**Reply by Navid200** [reply]
CareLink was added and often the individual who added it creates his own variants.
When I saw your first post and asked to see the status pages, I wanted to see if possibly it was a private release that worked and the official release did not.  if that was the case, I would have tagged the developer and asked him to look and advise.

But, you posted results of your compilations.  I did not expect that.  I have no experience with CarelInk.  I tag individuals who are involved and try to help move things forward.
Obviously, that did not work in this case and happily will stay out of this thread.

**Comment by jamorham** [comment]
This is quite possible, add the `@Keep` annotation above the top line of the class definition for data classes which are populated by deserialization/reflection and see if that helps.

**Reply by kukuchta** [reply, marked as answer]
Okay, proGuard was the case. To fix it in single place I just added:

`-keep class com.eveningoutpost.dexdrip.cgm.carelinkfollow.message.*`

at the end of app/proguard-rules.pro file, and that fixed Carelink notifications issue on default build variant (fastDebug).

**Reply by kukuchta** [reply]
How can we proceed from here? Should I prepare the PR?

**Reply by jamorham** [reply]
For a very simple change like this I can just add it

**Reply by kukuchta** [reply]
Would be great, thank you.
I've tested this by running both build versions for a while on the emulator. I didn't see any side effects, I dont think adding this rule will break anything.
